### ISSUE TITLE
Using `[AllowNull]` with field keyword should synthesize a nullable backing field when field is assigned a possibly null value

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#84720]
```csharp
class C
{
    private string? _s;

    [AllowNull]
    public string S
    {
        get => _s ?? string.Empty;
        set => _s = value;
    }

    [AllowNull]
    public string S2
    {
        get => field ?? string.Empty;
        // warning CS8601: Possible null reference assignment.
        // This warning is unexpected.
        set => field = value;
    }
}
```

The compiler could probably do the nullability analysis of 'field' assignment, and if **any** code path is assigning a possibly-null value, then the backing field is nullable. Otherwise, the backing field isn't nullable.

There is also an unexpected CS9264, tracked separately in #84719

Up to the compiler team to decide if both issues should be fixed together.

**Comment by Youssef1313** [comment]
Looks fixed. I was testing something on SharpLab which was very outdated. Not reproducible in https://lab.razor.fyi/

**Comment by RikkiGibson** [comment]
Most likely the auto nullability makes this just do the right thing. i.e. getter returns non-null value, even when field has maybe-null initial state, so we make the field nullable.
